powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/ParamManager.java

```diff
@@ -26,21 +31,46 @@
 
     private static final CacheManager cacheManager = new CacheManager();
     private static final TransformationManager transformationManager = new TransformationManager();
+    private static final ConcurrentHashMap<Class<? extends BaseProvider>, BaseProvider> providers = new ConcurrentHashMap<>();
 
     private static SecretsProvider secretsProvider;
     private static SSMProvider ssmProvider;
 
+    /**
+     * Get a concrete implementation of {@link BaseProvider}.<br/>
+     * You can specify {@link SecretsProvider} or {@link SSMProvider} or create your custom provider
+     * by extending {@link BaseProvider} if you need to integrate with a different parameter store.
+     * @return a {@link SecretsProvider}
+     */
+    public static <T extends BaseProvider> T getProvider(Class<T> providerClass) {
+        if (providerClass == null) {
+            throw new IllegalStateException("You cannot provide a null provider class.");
+        }
+        try {
+            if(!providers.containsKey(providerClass)) {
+                Constructor<T> constructor = providerClass.getDeclaredConstructor(CacheManager.class);
+                T provider = constructor.newInstance(cacheManager);
+                provider.setTransformationManager(transformationManager);
+                providers.put(providerClass, provider);
+            }
+            return (T) providers.get(providerClass);
+        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            throw new ProviderException(e);
+        }
+    }
+
     /**
      * Get a {@link SecretsProvider} with default {@link SecretsManagerClient}.<br/>
      * If you need to customize the region, or other part of the client, use {@link ParamManager#getSecretsProvider(SecretsManagerClient)} instead.
      * @return a {@link SecretsProvider}
      */
     public static SecretsProvider getSecretsProvider() {
-        if (secretsProvider == null) {
+        if (!providers.containsKey(SecretsProvider.class)) {
             secretsProvider = SecretsProvider.builder()
                                     .withCacheManager(cacheManager)
                                     .withTransformationManager(transformationManager)
                                     .build();
+            providers.put(SecretsProvider.class, secretsProvider);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **68 and 73**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **75**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.